### PR TITLE
feat: gracefully handle unsuccessful metrics post

### PR DESCRIPTION
**Pull request by ivarconr**
fixes: #405

### CODE REVIEW

File: src/metrics.ts [hunk 1]
```diff
@@ -180,107 +194,124 @@ export default class Metrics extends EventEmitter {
...
-      sdkVersion: this.sdkVersion,
-      strategies: this.strategies,
-      started: this.started,
-      interval: this.metricsInterval,
+      bucket,
     };
   }
 
-  getMetricsData(): Data {
+  private reAddBucket(bucket: Bucket) {
```
Comment (gardleopard) [inline src/metrics.ts:289]: I think reAddBucket could be called copyBucket, initializeBucket or something. reAddBucket naming is based on the context of when its called more than what it does.

**Review by gardleopard — APPROVED** [review]
LGTM

File: src/metrics.ts [hunk 2]
```diff
@@ -180,107 +194,124 @@ export default class Metrics extends EventEmitter {
...
-    }
+    this.increaseVariantCounter(name, variantName, 1);
 
     this.emit(UnleashEvents.CountVariant, name, variantName);
     return true;
   }
 
-  private bucketIsEmpty() {
-    if (!this.bucket) {
+  private increaseCounter(name: string, enabled: boolean, inc = 1): boolean {
```
Comment (sighphyre) [inline src/metrics.ts:243]: Do we do anything with the return value here or is this to keep in line with the other similar methods?

File: src/metrics.ts [hunk 3]
```diff
@@ -180,107 +194,124 @@ export default class Metrics extends EventEmitter {
...
     }
+    this.assertBucket(name);
+    this.bucket.toggles[name][enabled ? 'yes' : 'no'] += inc;
+    return true;
+  }
+
+  private increaseVariantCounter(name: string, variantName: string, inc = 1): boolean {
+    this.assertBucket(name);
+    if(this.bucket.toggles[name].variants[variantName]) {
+      this.bucket.toggles[name].variants[variantName]+=inc
```
Comment (sighphyre) [inline src/metrics.ts:255]: If you make any changes before merging, would you mind applying an autoformat here? It's not a big deal but it's nice

File: src/metrics.ts [hunk 4]
```diff
@@ -180,107 +194,124 @@ export default class Metrics extends EventEmitter {
...
-      this.bucket.stop = new Date();
-    }
   }
 
-  private getPayload(): Data {
-    this.closeBucket();
-    const payload = this.getMetricsData();
-    this.resetBucket();
-    return payload;
+  private resetBucket(): Bucket {
```
Comment (sighphyre) [inline src/metrics.ts:274]: My gut feel on this function is that it should either mutate state or return a value but not both, it becomes hard to reason about when you see the return value but it's also quietly mutating state behind the scenes

File: src/url-utils.ts [hunk 5]
```diff
@@ -1,4 +1,12 @@
-import { resolve } from 'url';
+export function resolve(from: string, to: string) {
```
Comment (sighphyre) [inline src/url-utils.ts:1]: What about `resolveUrl` instead for a name? When it's in the context of `url-utils`, it's easy to know what this does but once its imported and used in another file, it becomes hard to know what this does from the name without looking at its import location

**Review by sighphyre — APPROVED** [review]
LGTM, I've left a few thoughts but I don't think anything is deal breaking

File: src/metrics.ts (hunk 2 above)
Comment (ivarconr) [reply src/metrics.ts:243]: lol. these are way old. no real meaning, but was used in tests. will clean up.

Comment (ivarconr) [reply src/metrics.ts:243]: fixed

File: src/metrics.ts (hunk 3 above)
Comment (ivarconr) [reply src/metrics.ts:255]: done

File: src/url-utils.ts (hunk 5 above)
Comment (ivarconr) [reply src/url-utils.ts:1]: yes. this was a case of me just coping the old deprecated method from url. Agree that resolveUrl is a better name. fixed!

File: src/metrics.ts (hunk 4 above)
Comment (ivarconr) [reply src/metrics.ts:277]: yes, thus ended up as a foot gun. fixing.

**Pull request merged**